pkg/operations/graph.go

```diff
@@ -0,0 +1,252 @@
+	return nil
+}
+
+func (gp *GraphProg) Insert(what string) error {
+
+	// Derive and check some arguments:
+	if gp.DocConfig.KeySize < 1 || gp.DocConfig.KeySize > 64 {
+		gp.DocConfig.KeySize = 64
+	}
+
+	var gg graphgen.GraphGenerator
+
+	switch gp.GraphType {
+	case "cyclic":
+		gg = graphgen.NewCyclicGraph(gp.GraphSize, int(gp.DocConfig.KeySize), gp.VertexCollName)
+	default:
+		return fmt.Errorf("Unknown graph type: %s", gp.GraphType)
+	}
+
+	var err error
+	var numberDocs int64
+	var numberBatches int64
+	var ch chan *datagen.Doc
+	if what == "vertices" {
+		// Number of batches to put into the collection:
+		numberDocs = gg.NumberVertices()
+		numberBatches = (numberDocs + gp.BatchSize - 1) / gp.BatchSize
+		ch = gg.VertexChannel()
+	} else if what == "edges" {
+		numberDocs = gg.NumberEdges()
+		numberBatches = (numberDocs + gp.BatchSize - 1) / gp.BatchSize
+		ch = gg.EdgeChannel()
+	} else {
+		return fmt.Errorf("unknown insert task %s", what)
+	}
+	gp.DocConfig.SizePerDoc = gp.DocConfig.Size / numberDocs
+
+	config.OutputMutex.Lock()
+	fmt.Printf("\ngraph (%s): Will write %d batches of %d docs across %d goroutines...\n\n", what, numberBatches, gp.BatchSize, gp.Parallelism)
+	config.OutputMutex.Unlock()
+
+	err = database.RunParallel(gp.Parallelism, gp.StartDelay, "graph",
+		func(id int64) error {
+			// Let's use our own private client and connection here:
+			cl, err := config.MakeClient()
+			if err != nil {
+				return fmt.Errorf("Can not make client: %v", err)
+			}
+			db, err := cl.Database(context.Background(), gp.Database)
+			if err != nil {
+				return fmt.Errorf("Can not get database: %s", gp.Database)
+			}
+
+			var coll driver.Collection
+			if what == "vertices" {
+				coll, err = db.Collection(nil, gp.VertexCollName)
+			} else {
+				coll, err = db.Collection(nil, gp.EdgeCollName)
+			}
+			if err != nil {
+				config.OutputMutex.Lock()
+				fmt.Printf("graph write %s: could not open `%s` collection: %v\n", what, coll.Name(), err)
+				config.OutputMutex.Unlock()
+				return err
+			}
+
+			docs := make([]datagen.Doc, 0, gp.BatchSize)
+			times := make([]time.Duration, 0, numberBatches/gp.Parallelism)
+			cyclestart := time.Now()
+			last100start := cyclestart
+
+			// It is crucial that every go routine has its own random source,
+			// otherwise we create a lot of contention.
+			source := rand.New(rand.NewSource(int64(id) + rand.Int63()))
+
+			i := int64(0)
+			nrDocs := int64(0)
+			for done := false; !done; {
+				start := time.Now()
+				for j := int64(1); j <= gp.BatchSize; j++ {
+					doc, ok := <-ch
+					if !ok {
+						done = true
+						break
+					}
+					doc.FillData(&gp.DocConfig, source)
+					docs = append(docs, *doc)
+				}
+				if len(docs) > 0 {
+					ctx, cancel := context.WithTimeout(driver.WithOverwriteMode(context.Background(), driver.OverwriteModeIgnore), time.Hour)
+					_, _, err := coll.CreateDocuments(ctx, docs)
```

Comment:
For inserting documents, doesn't it need to say something about edges or vertices? Otherwise wouldn't it create a normal document without `_from` and `_to` and such graph specifics?

Reply:
I do not think so. When I find the collection, it notices that it is an edge collection. If I then call `coll.CreateDocuments` it chooses the method which is adequate for an edge collection. This is similar to inheritance and virtual methods. Here, `Collection` is an interface type, `coll` is of type `Collection` (which is a pointer, since it is an interface) and then, since both vertex collections and edge collections implement the interface `Collection`, the right method is chosen. This is similar to a vtable in C++, here there is a vtable for each type implementing an interface type.